src/standalone/Soil/boundary_conditions.jl

```diff
@@ -909,23 +912,99 @@ function soil_boundary_fluxes!(
+    liquid_influx = compute_liquid_influx(p, model, prognostic_land_components)
+    # This partitions the influx into runoff and infiltration
+    update_infiltration_water_flux!(p, bc.runoff, liquid_influx, Y, t, model)
+    # This computes the energy of the infiltrating water
+    infiltration_energy_flux = compute_infiltration_energy_flux(
+        p,
+        bc.runoff,
+        bc.atmos,
+        prognostic_land_components,
+        liquid_influx,
+        model,
+        Y,
+        t,
+    )
+    # The actual boundary condition is a mix of liquid water infiltration and
+    # evaporation.
+    @. p.soil.top_bc.water =
+        p.soil.infiltration + p.soil.turbulent_fluxes.vapor_flux_liq
     @. p.soil.top_bc.heat =
-        p.soil.R_n + p.soil.turbulent_fluxes.lhf + p.soil.turbulent_fluxes.shf
+        p.soil.R_n +
+        p.soil.turbulent_fluxes.lhf +
+        p.soil.turbulent_fluxes.shf +
+        infiltration_energy_flux
     return nothing
 end
 
+"""
+   compute_liquid_influx(p,
+                         model,
+                         prognostic_land_components::Val{(:soil,)},
+    ) 
+
+Returns the liquid water volume flux at the surface of the soil; this
+will then be partitioned into surface runoff and infiltration:
+influx = infiltration - runoff (runoff > 0; infiltration and influx < 0,
+and infiltration <= influx).
+
+In a model without snow as a prognostic variable, the influx is simply
+the liquid precipitation as a volume flux.
+"""
+function compute_liquid_influx(
+    p,
+    model,
+    prognostic_land_components::Val{(:soil,)},
+)
+    return p.drivers.P_liq
+end
+
+"""
+    compute_infiltration_energy_flux(
+        p,
+        runoff,
+        atmos::PrescribedAtmosphere,
+        prognostic_land_components::Val{(:soil,)},
+        liquid_influx,
+        model::EnergyHydrology,
+        Y,
+        t,
+    )
+
+Computes the energy associated with infiltration of
+liquid water into the soil.
+
+If the source of the infiltration is purely liquid precipitation,
+we approximate the volumetric internal energy with the volumetric
+internal energy of liquid water at the air temperature at the surface.
+
+Future runoff parameterizations may require allowing for exfiltration of 
+liquid water from the soil, which would have a different energy. To accomodate
+this, the runoff is passed as an argument and this can be used for dispatch in the 
+future.
+
+Furthermore, in couple simulations, the energy flux may be provided directly,
+rather than having the land model compute it. To accomodate this, a new method
+can be defined which dispatches off of the `atmos` type.
+"""
+function compute_infiltration_energy_flux(
+    p,
+    runoff,
+    atmos::PrescribedAtmosphere,
+    prognostic_land_components::Val{(:soil,)},
+    liquid_influx,
+    model::EnergyHydrology,
+    Y,
+    t,
+)
+    earth_param_set = model.parameters.earth_param_set
+    return @lazy @. p.soil.infiltration * Soil.volumetric_internal_energy_liq(
```

Comment:
I'm not sure about the `@lazy @.` syntax. Maybe check with @dennisYatunin. Wouldn't `@.lazy()` do? And maybe it's better to leave the `lazy` on the outermost layer (i.e., not inside the function) in any case.